### ISSUE TITLE
🐛 Bug Report: KubernetesContainerRunner adds a `/` to the mount base path irrespective of the OS

### ISSUE THREAD

**Issue by SayakMukhopadhyay** [backstage/backstage#16276]
### 📜 Description

The `KubernetesControllerRunner` while validating its volume mount, checks if the mount path ends with `/` and ads it if it doesn't. This causes it to be have weirdly in Windows, especially since I am developing on it.
https://github.com/backstage/backstage/blob/1f2b2de3fe24746297efc6596f8d5192a45c60f2/packages/backend-common/src/util/KubernetesContainerRunner.ts#L110-L112
I am not sure why the check and addition is there in the first place. Maybe @matteosilv can shed some light on the reason since they were the one who submitted this feature.

### 👍 Expected behavior

I expect that the `mountBase.basePath` would be accepted unaltered. And if a slash is necessary it should be supplied based on the OS.

### 👎 Actual Behavior with Screenshots

The above change in the `mountBase.basePath` causes the `runContainer` to throw an error when I pass the same path for both the `mountBase.basePath` and `mountDirs`. Here are some snippets of configuration that I expect to work
```ts
const containerRunner = new KubernetesContainerRunner({
  kubeConfig,
  name: 'scaffolder',
  mountBase: {
    basePath: ctx.workspacePath,
    volumeName: 'workdir',
  },
  ...
});
...
await containerRunner.runContainer({
  ...
  mountDirs: {
    [ctx.workspacePath]: '/input',
  },
  ...
});
```

### 👟 Reproduction steps

Just setup an action that uses the `KubernetesControllerRunner` and configure it as above.

### 📃 Provide the context for the Bug.

I am creating a custom action which will run a gradle task on my templated files before publishing. My backstage app is deployed on k8s and as such I want to leverage k8s jobs to run the scaffolder action which uses an image of gradle.

### 🖥️ Your Environment

OS:   Windows_NT 10.0.22623 - win32/x64
node: v18.12.0
yarn: 1.22.19
cli:  0.22.1 (installed)
backstage:  1.10.1

Dependencies:
  @backstage/app-defaults                          1.1.0
  @backstage/backend-app-api                       0.3.1
  @backstage/backend-common                        0.18.1
  @backstage/backend-plugin-api                    0.3.1
  @backstage/backend-tasks                         0.4.2
  @backstage/catalog-client                        1.3.0
  @backstage/catalog-model                         1.1.5
  @backstage/cli-common                            0.1.11
  @backstage/cli                                   0.22.1
  @backstage/config-loader                         1.1.8
  @backstage/config                                1.0.6
  @backstage/core-app-api                          1.4.0
  @backstage/core-components                       0.12.3
  @backstage/core-plugin-api                       1.3.0
  @backstage/errors                                1.1.4
  @backstage/integration-aws-node                  0.1.1
  @backstage/integration-react                     1.1.9
  @backstage/integration                           1.4.2
  @backstage/plugin-api-docs                       0.8.14
  @backstage/plugin-app-backend                    0.3.41
  @backstage/plugin-auth-backend                   0.17.4
  @backstage/plugin-auth-node                      0.2.10
  @backstage/plugin-catalog-backend-module-github  0.2.4
  @backstage/plugin-catalog-backend                1.7.1
  @backstage/plugin-catalog-common                 1.0.10
  @backstage/plugin-catalog-graph                  0.2.26
  @backstage/plugin-catalog-import                 0.9.4
  @backstage/plugin-catalog-node                   1.3.2
  @backstage/plugin-catalog-react                  1.2.4
  @backstage/plugin-catalog                        1.7.2
  @backstage/plugin-events-node                    0.2.2
  @backstage/plugin-github-actions                 0.5.14
  @backstage/plugin-org                            0.6.4
  @backstage/plugin-permission-common              0.7.3
  @backstage/plugin-permission-node                0.7.4
  @backstage/plugin-permission-react               0.4.9
  @backstage/plugin-proxy-backend                  0.2.35
  @backstage/plugin-scaffolder-backend             1.10.1
  @backstage/plugin-techdocs                       1.4.3
  @backstage/plugin-user-settings                  0.6.2
  @backstage/release-manifests                     0.0.8
  @backstage/test-utils                            1.2.4
  @backstage/theme                                 0.2.16
  @backstage/types                                 1.0.2
  @backstage/version-bridge                        1.0.3

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by matteosilv** [comment]
Check is required because it then slices it and the subPath is supposed not to start with / and be relative. I did not test it on Windows. Can you be more specific on what the error is on windows?

**Comment by SayakMukhopadhyay** [comment]
I didn't notice the `slice`. Still, the issue I am having and might have:
https://github.com/backstage/backstage/blob/1f2b2de3fe24746297efc6596f8d5192a45c60f2/packages/backend-common/src/util/KubernetesContainerRunner.ts#L151
In the above `hostDir` is supposed to be a subdirectory of the `mountBase.basePath`. In Windows, `mountBase.basePath` would be like `C:\some\path\in\drive/` and `hostDir` could be `C:\some\path\in\drive\sub`. So, even though `hostDir` is a legit subdirectory of  `mountBase.basePath`, the above check would still fail.

Moreover, I am a bit confused on which path stands for what. I tried the scaffolder using the `DockerContainerRunner` and it works fine for me locally but wouldn't work on k8s due to a DInD situation. So, I decided to use `KubernetesContainerRunner` as I can use k8s both locally and on remote. But, now the various paths seems to mean different things.

In `DockerContainerRunner`, the config `mountDirs` accepts a map where the key pointed to the folder in the host and the value pointed to the mount point of that folder in the container.

The semantics seems to have changed in `KubernetesContainerRunner`. Am I on the right track here?

**Comment by matteosilv** [comment]
Yeah the semantics is a bit different with the kubernetescontainerrunner. I can use the right separator based on os, but I have no way to test it. If I do the edit, would you mind to test it, or I could check if there’s a way to test it via github workflows

**Comment by SayakMukhopadhyay** [comment]
Sure I can test it. But I still think that adding the separator could cause subdirectory check to fail if I am not careful with the `hostDir`. As you can see in the OP, I am using `ctx.workspacePath` directly as the `hostDir` and `ctx.workspacePath` doesn't have the trailing slash, at least not in Windows.

Rather, my suggestion would be to not add the trailing slash and instead do the `slice` either by `this.mountBase.basePath.length` or `this.mountBase.basePath.length - 1` depending on whether the user has entered the trailing slash or not. This is also closer to the need of the trailing slash and will ensure that other side effects won't happen due to the addition of the trailing slash.

**Comment by matteosilv** [comment]
@SayakMukhopadhyay sorry mate for taking too long to answer! 
- Yes I'm in the discord channel. We can chat!
- Yes I wrote this for the scaffolder. I needed to create a new container:run action to run an arbitrary action, but I was not able to do it with the DockerContainerRunner. What I do is I mount the same volume, specifically the backend workingDirectory in both the backend itself and the minions spawn by the Kubernetes container runner. The result is that every task has its subdirectory mounted in the pod running the job
